services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java

```diff
@@ -479,6 +482,25 @@ private void updateAuthenticationSession() {
         if (request.getCodeChallenge() != null) authenticationSession.setClientNote(OIDCLoginProtocol.CODE_CHALLENGE_PARAM, request.getCodeChallenge());
         if (request.getCodeChallengeMethod() != null) authenticationSession.setClientNote(OIDCLoginProtocol.CODE_CHALLENGE_METHOD_PARAM, request.getCodeChallengeMethod());
 
+        Map<String, Integer> acrLoaMap = AcrUtils.getAcrLoaMap(authenticationSession.getClient());
+        List<String> acrValues = AcrUtils.getRequiredAcrValues(request.getClaims());
+
+        if (acrValues.isEmpty()) {
+            acrValues = AcrUtils.getAcrValues(request.getClaims(), request.getAcr());
+        } else {
+            authenticationSession.setClientNote(Constants.FORCE_LEVEL_OF_AUTHENTICATION, "true");
+        }
+
+        acrValues.stream().mapToInt(acr -> {
+            try {
+                Integer loa = acrLoaMap.get(acr);
+                return loa == null ? Integer.parseInt(acr) : loa;
+            } catch (NumberFormatException e) {
+                // this is an unknown acr, we assume it is very high
+                return Constants.MAXIMUM_LOA;
```

Comment:
@cgeorgilakis I agree. It should fail and we can not assume the LoA. I think I understand the reasoning about setting MAX, but looks more correct to just fail because the requirement can not be met.

Reply:
@pedroigor In order to be clear. It should fail only if claims with essential='true' are all unknown. Otherwise, it should omit this claim value.

I do not know when it will fail. I believe that this is implementation decision.

Reply:
@cgeorgilakis @pedroigor Thanks for the review - we really appreciate your valuable feedback!

If the requested acr value is unknown, we return the MAXIMUM_LOA (maximum integer) and this results in an authentication error because the user can never reach this LoA. The authentication error only occurs if the claim is essential="true". Otherwise, we omit an unknown requested acr value. 
As already suspected, the error only occurs after the user has provided the credentials.
Additionally, if an authorization request contains more than one (known) acr value, the minimum of the known acr values is processed.

Sorry for the late reply. We are currently going live with one of our services and therefore our resources are quite blocked, so thanks in advance for you patience. Hence, we are not yet able to answer all comments, but we will try our best to do so in the next week. We will rebase this branch by tomorrow.

Reply:
@CorneliaLahnsteiner you are right about essential="true". This is the most important.

However, think about the following situation:

> Flow Configuration is same as you descipted in .md file( 1st -> username-password, 2nd -> otp). 
Admin have not set acr value in oidc client.
User is already login with level 1.
Client send only acr 'silver' without essential attribute in claim. Keycloak will set REQUESTED_LEVEL_OF_AUTHENTICATION to highest. So, user will ask to authedicate with 2nd LOA ( otp form) and will return 'silver' ( or 2) as acr.
Documentation says '  If the Claim is not Essential and a requested value cannot be provided, the Authorization Server SHOULD return the session's current acr as the value of the acr Claim.'

It would be great if keycloak in this situation do not ask user to authedicate with otp and return 1 as acr.
However, this issue is little importance. The most important for our team is Keycloak being able to support step authedication based on this PR. Default behaviour should be written in Keycloak documentation.

Reply:
> @pedroigor In order to be clear. It should fail only if claims with essential='true' are all unknown. Otherwise, it should omit this claim value.

Looks like that the changes here fail even for voluntary claims? I think it makes more sense because in either case (essential or voluntary) they are unknown.

If the current behavior is like that. It is OK for me.

Reply:
>It would be great if keycloak in this situation do not ask user to authedicate with otp and return 1 as acr.
However, this issue is little importance. The most important for our team is Keycloak being able to support step authedication based on this PR. Default behaviour should be written in Keycloak documentation.

It works for me. I'm missing why a client would want to send `silver`, as in this case, as a voluntary claim if the AS should ignore and send the current `acr`.

Reply:
If essential="true" and all ACRs are unknown, we could fail directly, but it is not clear to me how to implement it with a useful error page in this method.
In the case of an optional ACR where all provided values are unknown, the server will process all authenticators to reach the MAXIMUM_LOA. But instead to fail, it will return the final reached acr value. The ClientNote FORCE_LEVEL_OF_AUTHENTICATION decides in the end wheter to fail or return the final ACR.

Reply:
@cgeorgilakis Once the code review is complete and we have a final draft of the code, we will of course provide detailed documentation on Step up Authentication.

Reply:
> If essential="true" and all ACRs are unknown, we could fail directly, but it is not clear to me how to implement it with a useful error page in this method. In the case of an optional ACR where all provided values are unknown, the server will process all authenticators to reach the MAXIMUM_LOA. But instead to fail, it will return the final reached acr value. The ClientNote FORCE_LEVEL_OF_AUTHENTICATION decides in the end wheter to fail or return the final ACR.

Thanks for all work on this needed future.
However, implementation for optional ACR not defined in Keycloak client is not following OIDC Specification proposal.

> If the Claim is not Essential and a requested value cannot be provided, the Authorization Server SHOULD return the session's current acr as the value of the acr Claim.

I believe that change is easy in order to be allign with OIDC specification. 
Moreover, it Keycloak documentation should provide information about unknown acr.

Reply:
@cgeorgilakis We fixed this behavior for a not essential and unknown acr claim: https://github.com/keycloak/keycloak/pull/7897/commits/4081d8151d1c023be1366e563d7487bed2e19b52#diff-20c882d1eedd43d45e8336221b029e174ff2dc5acf324092b7acdc1cebb98936R310

In detail: Until now, an unknown acr claim was always treated with a maximum LoA, which means that one had to authenticate with username+password and OTP - no matter if the acr claim is essential or not. We have adjusted this behaviour so that in the case of an essential (unknown) acr claim, the maximum LoA is maintained, but in the case of a non-essential (unknown) acr claim, the user is only asked to present one authentication factor instead of both (if he has no session yet). Consequently, the access token and ID token will include the acr of 1 after an successful authentication.

Reply:
@cgeorgilakis I will also try my best to adapt the documentation to highlight the different possibilities (essential vs. non-essential claims and unknown claims).

Reply:
Nice!